[master] Fix 2 pkgrepo.absent bugs for apt-based distros

terminalmage opened the pull request:
### What does this PR do?

Resolves 2 pkgrepo.absent bugs

### What issues does this PR fix or reference?
Fixes: #64286

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Ch3LL commented on changelog/64286.fixed.md:
```diff
@@ -0,0 +1 @@
+Fix 2 pkgrepo.absent bugs for apt-based distros
```
Can  you clarify what bugs this is resolving.

terminalmage replied:
The reason I left it this way is because I figured the changelog would link to the issue, which would have the further explanation. I can try to summarize this though. I'll update this when I get tests written, hopefully by the end of the week.

s0undt3ch replied:
As an FYI, the issue(s) do get linked on the rendered changelog.
A good summary however goes a long way.

Thank You @terminalmage for the issue and the PR to fix it

Ch3LL reviewed (changes requested):
This will also require test coverage. Thanks for this :)

terminalmage committed:
Revise changelog

terminalmage committed:
Remove unused fixture argument

terminalmage commented:
@Ch3LL It turned out that my fix had downstream effects that caused problems with `pkgrepo.managed`, so I had to revise the fix. In the process I exposed one of the private functions from `aptpkg.py` so it could be invoked from within the `pkgrepo` state module.

Going to bed now, will check on tests in the morning :slightly_smiling_face:

terminalmage committed:
Remove incorrect requirement, and add additional comments to test docstring

terminalmage committed:
Update aptpkg unit tests to reflect rename of _split_repo_str

terminalmage commented:
The pkgrepo tests are passing. [This RPM build test](https://github.com/saltstack/salt/actions/runs/5068371426/jobs/9103224570?pr=64287) is failing though, and there appear to be no details there so I don't know what caused it.

It also looks like some CI runners might be down because several of the docs tests are stalled waiting for a free runner.

Ch3LL commented on salt/modules/aptpkg.py:
```diff
@@ -1756,23 +1756,32 @@ def _get_opts(line):
+def apt_source_entry(repo):
     """
-    Return APT source entry as a tuple.
+    .. versionadded:: 3006.2
```
If this is a new feature this would need to be 3007.0, not 3006.2 and a changelog would also need to be added for adding this new function in addition to the one you current have added. ( a changelog with a `.added.md` extension)

Ch3LL replied:
But a user could not have previously run this function, and now they can, so that part is a new feature. We could separate the PRs out:
1. push a PR to the `3006.x` branch keeping the private function with the changes, without the `versionadded`.
2. keep this PR pushed to master, but updating the versionadded to 3007.0.

s0undt3ch replied:
We could import the private function and namespace it into the state module?
That way we would be coherent?

Ch3LL replied:
@terminalmage I can certainly see your point here. Let me discuss with the team and I'll get back to you here.

terminalmage replied:
I do not understand what is incoherent about this PR.

I also do not understand how importing and invoking a private function is more coherent than checking if the function is in the loader, as salt/states/pkgrepo.py already does [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkgrepo.py#L361), [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkgrepo.py#L688), and [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkgrepo.py#L136).

s0undt3ch replied:
That's not the point, the point is exposing(adding) new functionality in a point release, a feature addition.
Yes, the import is more cumbersome, and would have to be undo for the `master` branch to do it your way...

Anyway, let's see what the rest of the team has to say about this.

s0undt3ch replied:
The import should be done after the Debian family check

terminalmage replied:
This solution would work, sure, but what about cases where people are manually specifying the provider (via the [providers](https://docs.saltproject.io/en/latest/ref/configuration/minion.html#providers) minion config option)? This is one of the reasons we have historically tried to avoid grains checks in the states, opting instead for `__salt__` membership checks.

Sometimes core grains become out-of-date, when new distros arise (or new versions of distros do things differently), leading to lapses in Salt's ability to properly detect OS. In those cases, `providers` is a good way to get things working while Salt catches up to these changes. The proposed alternative would break `pkgrepo.absent` for these minions.

I can think of 4 solutions (please suggest alternatives if any come to mind):


1. Implement the proposed alternative solution from [this comment](https://github.com/saltstack/salt/pull/64287#discussion_r1212065476).
   - This would break minions which are manually set to use aptpkg as the `pkg` provider, as mentioned above.
2. Implement the proposed alternative solution, but also check `__opts__` to see if `providers` is in use for `pkg`.
   - `__opts__.get('providers', {}).get('pkg') == 'aptpkg'`
   - While this would technically work, IMO it defeats the purpose of using `providers`.
   - Furthermore, it breaks with established precedent of using loader membership to run code that is platform-specific. Some examples:
     1. `pkg.check_db` only exists for Gentoo Linux (and derivatives which use `portage`). The function serves little purpose for CLI usage, and is arguably there simply for programmatic purpose. Yet, `pkg` states [check the loader](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L946-L962) to see whether or not that function is there to be run.
     2. `pkg.list_provides` only exists for SUSE, yet Salt checks the loader [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L565-L574) and [here](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L2030-L2034) to see whether or not this function is available to be run.
     4. The `lowpkg` virtual module only exists for systems that use DEB or RPM packages. Yet states [check if `lowpkg.bin_pkg_info` is present](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L756) to see whether or not it can parse package metadata for states which use `sources` to install package files without using a DEB/RPM repository. The functions in `lowpkg` don't serve a meaningful purpose on the CLI and are used for programmatic reasons.
     5. `pkg.verify` only exists for `yum` and `zypper`, yet states [check the loader](https://github.com/saltstack/salt/blob/c33b562/salt/states/pkg.py#L1747) to make sure that package verification is available.
3. Keep the code as-is, but target 3007.0. (docstrings would need to be updated). This decision would be my least favorite, since it would be a conscious decision to 2 allow easily-fixable bugs to remain in the codebase.
4. Make no changes to the pull request.

Ch3LL replied:
We would like to go forward with option 2. The issue isn't adding the ability to check to see if the function exists in the loader, the issue is we are making a function available to users that does not add value, but only for programmatic purposes. If this is something that is already being done in the code today, we should consider that technical debt that would need to be fixed.

terminalmage replied:
This has been completed.

Ch3LL commented:
@garethgreenaway mind reviewing this as well? I know you where in this code recently fixing a similar issue with trailing `/`

Ch3LL commented on salt/states/pkgrepo.py:
```diff
@@ -702,6 +708,12 @@ def absent(name, **kwargs):
         return ret
 
+    if repo and "pkg.apt_source_entry" in __salt__:
+        if set(__salt__["pkg.apt_source_entry"](stripname)["architectures"]) != set(
```
Note: I have not spoke to the team yet, but will today. I just had a follow up thought after yesterday. Could this logic be added to the `get_repo` function instead? Right here specifically https://github.com/saltstack/salt/blob/master/salt/modules/aptpkg.py#L2008  we could check if arch exists and compare the two.

terminalmage replied:
No, I tried this to begin with and doing so breaks `pkgrepo.managed`.

Ch3LL replied:
K, thanks for confirming.

terminalmage committed:
Do grains check and call imported utility function

Ch3LL commented:
@terminalmage looks like there are some related test failures

terminalmage commented:
@Ch3LL I don't know what is causing those errors. These tests pass on the commit before I added the ill-advised grains and opts checks, and adding that commit somehow breaks invoking an unrelated import (`salt.utils.pkg.deb.strip_uri`), even though A) `salt.utils.pkg.deb` is being imported in pkgrepo.py, and B) the failure comes _before_ the code I added.

This seems to be something related to the test suite setup. I'm going to need some help troubleshooting this.

terminalmage committed:
Fix 2 pkgrepo.absent bugs for apt-based distros

Commit:
Merge branch 'master' into pkgrepo-absent-bugs

terminalmage commented:
Sure, thanks!
